salt/modules/sysfs.py

```diff
@@ -0,0 +1,212 @@
+.. versionadded:: Boron
+'''
+# Import python libs
+from __future__ import absolute_import
+import logging
+import os
+import stat
+
+# Import external libs
+import salt.ext.six as six
+
+# Import salt libs
+import salt.utils
+
+log = logging.getLogger(__name__)
+
+
+def __virtual__():
+    '''
+    Only work on Linux
+    '''
+    return salt.utils.is_linux()
+
+
+def attr(key, value=None):
+    '''
+    Access/write a SysFS attribute.
+    If the attribute is a symlink, it's destination is returned
+
+    :return: value or bool
+
+    CLI example:
+     .. code-block:: bash
+
+        salt '*' sysfs.attr block/sda/queue/logical_block_size
+    '''
+    key = target(key)
+
+    if key is False:
+        return False
+    elif os.path.isdir(key):
+        return key
+    elif value is not None:
+        return write(key, value)
+    else:
+        return read(key)
+
+
+def write(key, value):
+    '''
+    Write a SysFS attribute/action
+
+    CLI example:
+     .. code-block:: bash
+
+        salt '*' sysfs.write devices/system/cpu/cpu0/cpufreq/scaling_governor 'performance'
+    '''
+    try:
+        key = target(key)
+        log.trace('Writing {0} to {1}'.format(value, key))
+        with open(key, 'w') as twriter:
+            twriter.write('{0}\n'.format(value))
+            return True
+    except:  # pylint: disable=bare-except
+        return False
+
+
+def read(key, root=None):
+    '''
+    Read from SysFS
+    :param key: file or path in SysFS; if key is a list then root will be prefixed on each key
+    :return: the full (tree of) SysFS attributes under key
+
+    CLI example:
+     .. code-block:: bash
+
+        salt '*' sysfs.read class/net/em1/statistics
+    '''
+
+    if not isinstance(key, six.string_types):
+        res = {}
+        for akey in key:
+            ares = read(os.path.join(root, akey))
+            if ares is not False:
+                res[akey] = ares
+        return res
+
+    key = target(os.path.join(root, key))
+    if key is False:
+        return False
+    elif os.path.isdir(key):
+        keys = interfaces(key)
+        result = {}
+        for subkey in keys['r'] + keys['rw']:
+            subval = read(os.path.join(key, subkey))
+            if subval is not False:
+                subkeys = subkey.split('/')
+                subkey = subkeys.pop()
+                subresult = result
+                if len(subkeys):
+                    for skey in subkeys:
+                        if skey not in subresult:
+                            subresult[skey] = {}
+                        subresult = subresult[skey]
+                subresult[subkey] = subval
+        return result
+    else:
+        try:
+            log.trace('Reading {0}...'.format(key))
+
+            # Certain things in SysFS are pipes 'n such.
+            # This opens it non-blocking, which prevents indefinite blocking
+            with os.fdopen(os.open(key, os.O_RDONLY | os.O_NONBLOCK)) as treader:
```

Comment:
Well I'd like to do that, but I don't think it's possible?

Take a good look at the admittedly somewhat exotic construct here; 
I'm using `os.open()` and passing that on to `os.fdopen()` rather than `open()` here for the very very specific `os.O_NONBLOCK` option. 

IIRC I did that because that was the only way to get that NONBLOCK option in the opener; `open()`, and therefore `salt.utils.fopen()` only supports 'mode', but I need that flag.

And we *need* that non-block option because there's some very exotic stuff in SysFS (pipes & sockets & such) 

Put differently; aside from the obvious DRY args I see that the main reason this util exists is to set the CLOEXEC option.

I haven't tested this yet, but I think there's a os.O_CLOEXEC which I could just add to get to the same point.
Secondly, if that turns out to be true, wouldn't that be a nice fix for salt.utils.fopen()? The fcntl wrap and race condition mentioned in the comments would both be replaced for a construct very similar to the one here.

I'm quite willing to patch both if that is the preferred course of action ;) 

@cachedout, what do you think?